File: src/CmdRefText.cpp
// CmdRefText.cpp - This file is part of QPlot

/* QPlot - Publication quality 2D graphs with dual coordinate systems
   Copyright (C) 2014  Daniel Wagenaar
  
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.
  
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
  
   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

// CmdReftext.C

#include "CmdRefText.h"
#include <QFontMetricsF>
#include "Rotate.h"
#include <QDebug>

static CBuilder<CmdRefText> cbRefText("reftext");

bool CmdRefText::usage() {
  return error("Usage: reftext [string]");
}

bool CmdRefText::parse(Statement const &s) {
  if (s.length()==2 && s[1].typ==Token::STRING)
    return true;
  else if (s.length()==1)
    return true;
  else
    return usage();
}

void CmdRefText::render(Statement const &s, Figure &f, bool) {
  if (s.length()==1)
    f.setRefText("");
  else
    f.setRefText(s[1].str);
}

File: python/qplot_backend/cmdreftext.py
# cmdreftext.py - This file is part of QPlot

# QPlot - Publication quality 2D graphs with dual coordinate systems
# Copyright (C) 2014  Daniel Wagenaar
# (same GPL licence as original)

from __future__ import annotations

from typing import TYPE_CHECKING

from .command import Command
from .token_ import Token

if TYPE_CHECKING:
    from .statement import Statement
    from .figure import Figure


@Command.register("reftext")
class CmdRefText(Command):
    """Set the reference text string used by CmdText for alignment.

    Syntax:
        reftext             (clears the reference text)
        reftext 'string'    (sets the reference text)
        reftext '+string'   (appends to the reference text)
    """

    def _usage(self) -> bool:
        return self.error("Usage: reftext [string]")

    def parse(self, s: Statement) -> bool:
        if len(s) == 1:
            return True
        if len(s) == 2 and s[1].typ == Token.STRING:
            return True
        return self._usage()

    def render(self, s: Statement, f: Figure, dryrun: bool) -> None:
        if len(s) == 1:
            f.set_ref_text("")
        elif s[1].str.startswith("+"):
            f.set_ref_text(f.ref_text() + s[1].str[1:])
        else:
            f.set_ref_text(s[1].str)
